File: public/ps/PS2/GraphADT/GraphADT/Edge.m
#import "Edge.h"

@implementation Edge

@synthesize source;
@synthesize destination;
@synthesize weight;

// Checks that the representation invariant holds.
-(void)checkRep{
    
	if (self.source == nil ){
		[NSException raise:@"Edge rep error" format:
		 @"Source of an Edge must be a valid Node object"];
	}
    
    if (self.destination == nil ){
		[NSException raise:@"Edge rep error" format:
		 @"Source of an Edge must be a valid Node object"];
	}
    
    if (self.weight == nil || ([self.weight compare:[NSNumber numberWithInt:0]] == NSOrderedAscending)){
		[NSException raise:@"Edge rep error" format:
		 @"Weight of an Edge must be a valid NSNumber greater than 0"];
	}
}

-(id)initWithSource:(Node*)s andDestination:(Node*)d andWeight:(NSNumber*)w {
    // MODIFIES: self
    // EFFECTS: returns a new Edge object e with e.source = s,
    //          e.destination = d and e.weight = w
    
    self = [super init];
    if (self) {
        source = s;
        destination = d;
        weight = w;
    }
    [self checkRep];
    return self;
}

-(id)initWithSource:(Node *)s andDestination:(Node *)d {
    // MODIFIES: self
    // EFFECTS: returns a new Edge object e with e.source = s,
    //          e.destination = d and e.weight = 1
    
    return [self initWithSource:s andDestination:d andWeight:[NSNumber numberWithInt:1]];
}

+(id)edgeWithSource:(Node*)s andDestination:(Node*)d andWeight:(NSNumber*)w {
    // EFFECTS: Factory method that creates a new Edge object e with
    //          e.source = s, e.destination = d and e.weight = w
    
    return [[Edge alloc] initWithSource:s andDestination:d andWeight:w];
}

+(id)edgeWithSource:(Node *)s andDestination:(Node *)d {
    // EFFECTS: Factory method that creates a new Edge object e with
    //          e.source = s, e.destination = d and e.weight = 1
    
    return [Edge edgeWithSource:s andDestination:d andWeight:[NSNumber numberWithInt:1]];
}

-(BOOL)isEqual:(id)object {
    // REQUIRES: self != nil, object != nil
    // EFFECTS: override NSObject's isEqual method.
    //          returns YES if object represents the same edge as self
    
    if([object isKindOfClass:[Edge class]]){
        // isKindOfClass returns true if object can be considered as an
        // instance of the argument class. Each class has a method
        // called "class", which returns the class object. so [Edge
        // class] returns the class object of Edge class
        
        Edge *e = (Edge*)object;
        return ([self.source isEqual:e.source] && [self.destination isEqual:e.destination] && [self.weight isEqualToNumber:e.weight]);
    }
    else {
        return NO;
    }
}

-(NSUInteger) hash {
    // REQUIRES: self != nil
    //  This is a private method that overrides the hash method of NSObject
    return [self.source hash];
}

@end

File: public/ps/PS2/GraphADT/GraphADT/Edge.swift
/*
    `Edge` represents an edge in a graph. An `Edge` is associated with
    a source Node, a destination Node and a non-negative cost (or weight). 
    This means that `Edge` is a directed edge from the source to 
    the destination.

    The representation invariants:
        - The weight is non-negative

    Similar to `Node`, `Edge` is a generic type with a type parameter
    `T` that is the type of a node's label.
*/


struct Edge<T: Equatable> {
    typealias N = Node<T>
    
    var source: N
    var destination: N
    var weight = 1.0
    
    //  Construct an edge from `source` to `destination` with the
    //  default cost of 1.0.
    init(source: N, destination: N) {
        self.source = source
        self.destination = destination
        _checkRep()
    }
    
    //  Construct an edge from `source` to `destination` with the
    //  cost of `weight`.
    init(source: N, destination: N, weight: Double) {
        self.source = source
        self.destination = destination
        self.weight = weight
        _checkRep()
    }
    
    //  Return an edge in the opposite direction with the same cost.
    func reverse() -> Edge<T> {
        return Edge(source: destination, destination: source, weight: weight)
    }
    
    //  Check the representation invariants.
    private func _checkRep() {
        assert(weight >= 0, "Edge's weight cannot be negative.")
    }
}


extension Edge : Equatable {
}


//  Return true if `lhs` edge is equal to `rhs` edge.
func ==<Label>(lhs: Edge<Label>, rhs: Edge<Label>) -> Bool {
    return lhs.source == rhs.source && lhs.destination == rhs.destination && lhs.weight == rhs.weight
}
